src/Cms/ModelLock.php

```diff
@@ -0,0 +1,319 @@
+<?php
+
+namespace Kirby\Cms;
+
+use Kirby\Data\Data;
+use Kirby\Toolkit\F;
+
+use Throwable;
+use Kirby\Exception\PermissionException;
+
+/**
+ * Takes care of content lock and unlock information
+ *
+ * @package   Kirby Cms
+ * @author    Nico Hoffmann <nico@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier GmbH
+ * @license   https://getkirby.com/license
+ */
+class ModelLock
+{
+
+    /**
+     * Lock data
+     *
+     * @var array
+     */
+    protected $data;
+
+    /**
+     * The model
+     *
+     * @var ModelWithContent
+     */
+    protected $model;
+
+    /**
+     * @param ModelWithContent $model
+     */
+    public function __construct($model)
+    {
+        $this->model = $model;
+        $this->data  = $this->read();
```

Comment:
Reading the lock file for each model will lead to issues if the lock of multiple models in the same directory (e.g. the page and a file or multiple files) is modified in the same request as each model will load the lock file once and then write it back, overwriting the lock changes from other models.

I think the most robust implementation would be a global `$kirby->locks()` collection that acts as a proxy:

- `$kirby->locks()->get(ModelWithContent $model)` would return the data array for a model. So all the model ID and lock file reading stuff would get handled by this (determine the correct lock file based on the model, open it, cache the contents of that lock file in an instance property, extract the array of the model by ID, return it).
- `$kirby->locks()->set(ModelWithContent $model, array $data)` would update the cache in the instance property and then write the lock file again.
- The `ModelLock` class could get slimmed down a lot because it wouldn't need to handle all the low-level file operations and model ID stuff. In the constructor, it would just `get()` its data with `$model->kirby()->locks()->get($model)` and on each change it would call `$this->kirby()->locks()->set($this->model(), $this->data)` (where `$this->data` no longer contains all lock file contents, but one level fewer: just the data of the current model).

I hope my explanation was good enough. If not, please let me know. :)

Reply:
Wouldn’t this increase the likelihood of race conditions as they’re now affected by different users editing any model at the same time while currently it’s only while editing models from the same content folder?

Reply:
There would still be a lock file per folder, so no difference there. Those different lock files would just be managed by a global class.

Keeping this global would also make it a lot easier to add FS locking later.

Reply:
Then I don’t understand the advantage yet. The Panel send individual requests per lock. So whether this is handled in the App class or in the Model class, it would only handle the data of one model since each API requests initialized Kirby for itself.

Reply:
For API requests yes, but if locks are modified manually in a script it will break.

Reply:
✅